core/src/main/java/org/apache/iceberg/taskio/BaseTaskWriter.java

```diff
@@ -0,0 +1,141 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg.taskio;
+
+import java.io.IOException;
+import java.util.Collections;
+import java.util.List;
+import java.util.function.Supplier;
+import org.apache.iceberg.DataFile;
+import org.apache.iceberg.DataFiles;
+import org.apache.iceberg.FileFormat;
+import org.apache.iceberg.Metrics;
+import org.apache.iceberg.PartitionKey;
+import org.apache.iceberg.PartitionSpec;
+import org.apache.iceberg.encryption.EncryptedOutputFile;
+import org.apache.iceberg.io.FileAppender;
+import org.apache.iceberg.io.FileIO;
+import org.apache.iceberg.relocated.com.google.common.collect.ImmutableList;
+import org.apache.iceberg.relocated.com.google.common.collect.Lists;
+import org.apache.iceberg.util.Tasks;
+
+abstract class BaseTaskWriter<T> implements TaskWriter<T> {
+  protected static final int ROWS_DIVISOR = 1000;
+
+  private final List<DataFile> completedFiles = Lists.newArrayList();
+  private final PartitionSpec spec;
+  private final FileFormat format;
+  private final FileAppenderFactory<T> appenderFactory;
+  private final OutputFileFactory fileFactory;
+  private final FileIO io;
+  private final long targetFileSize;
+
+  protected BaseTaskWriter(PartitionSpec spec, FileFormat format, FileAppenderFactory<T> appenderFactory,
+                           OutputFileFactory fileFactory, FileIO io, long targetFileSize) {
+    this.spec = spec;
+    this.format = format;
+    this.appenderFactory = appenderFactory;
+    this.fileFactory = fileFactory;
+    this.io = io;
+    this.targetFileSize = targetFileSize;
+  }
+
+  @Override
+  public void abort() throws IOException {
+    close();
+
+    // clean up files created by this writer
+    Tasks.foreach(completedFiles)
+        .throwFailureWhenFinished()
+        .noRetry()
+        .run(file -> io.deleteFile(file.path().toString()));
+  }
+
+  @Override
+  public List<DataFile> pollCompleteFiles() {
```

Comment:
I don't think it's a good idea to have a `poll` method like this one because it leaks critical state (`completedFiles`) and creates an opportunity for threading issues between `write` and `pollCompleteFiles`.

Instead, I think the base implementation should use a push model, where each file is released as it is closed.

```java
  /**
   * Called when a data file is completed and no longer needed by the writer.
   */
  protected abstract void completedFile(DataFile file);
```

Then `closeCurrent` would call `completedFile(dataFile)` and the implementation of `completedFile` would handle it from there.

Reply:
I read the `BaseWriter` code again and got the difference here.  For spark streaming writer,  once we did a `commit` ,  then we will create another new  streaming writer to write the future records,  so we don't need a method like `pollCompleteFiles()`  to poll the newly added `DataFile` continusely.  In the current iceberg flink writer implementation,  I will use the same TaskWriter to write record even if  a checkpoint happen,  so I designed the `pollCompleteFiles`  to fetch all completed data files incrementally.    I think it's design difference,  the state leaks issues and threading issues as you said,  it's not a problem in current version but I agree that it's easy to get into those issues if others did not handle it carefully.   I can align with the current spark design.